### PR TITLE
Add GEFS Ocean Perturbation Options

### LINKED ISSUE #2403: Put historical co2 data in the same place?
### What new functionality do you need?

I wonder if it is possible to move 

global_co2historicaldata_* that sits in this directory:

/scratch1/NCEPDEV/global/glopara/fix/am/20220805/

to this directory

/scratch1/NCEPDEV/global/glopara/fix/am/20220805/fix_co2_proj

Since this path is hard-coded is forecast_postdet as:
for file in $(ls "${FIXam}/fix_co2_proj/global_co2historicaldata"*) ; do

This way the same workflow can be used for cases older than 2009.

### What are the requirements for the new functionality?

moving files from one directory to another

### Acceptance Criteria

Files moved. 

### Suggest a solution (optional)

See above.

### PR DESCRIPTION
**Pull request by NeilBarton-NOAA**
# Description

This PR adds the MOM6 ocean perturbation options when running GEFS. This is needed for GEFS reforecasts and GEFS operational forecasts. 

This PR continues to address the below issues
https://github.com/NOAA-EMC/global-workflow/issues/1720
https://github.com/NOAA-EMC/global-workflow/issues/1921

Fixes #2403 

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

A successfully two member ensemble ran on hera using ICs at /scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/GW_TEST

# Checklist
- [X] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: ush/forecast_postdet.sh [hunk 1]
```diff
@@ -693,6 +693,11 @@ MOM6_postdet() {
       ${NLN} "${COM_OCEAN_ANALYSIS}/${RUN}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/mom6_increment.nc"
   fi
 
+  # GEFS perturbations
+  if (( 0${MEMDIR:3} > 0 )) && [[ "${ODA_INCUPD:-False}" == "True" ]]; then
+     ${NLN} "${COM_OCEAN_RESTART_PREV}/${sPDY}.${scyc}0000.mom6_increment.nc" "${DATA}/INPUT/mom6_increment.nc"
+  fi
+
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: We may need to wrap this in a `if [[ $RUN} == "gefs" ]]` block to ensure it does not interfere with the GFS ensemble.  That component is not ready yet, so we may just add a `TODO` for now and leave it at that.

File: ush/forecast_postdet.sh [hunk 2]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
-      ISEED_OCNSPPT=$(( (current_cycle*1000 + MEMBER*10 + 6) % 2147483647 ))
-      ISEED_EPBL=$(( (current_cycle*1000 + MEMBER*10 + 7) % 2147483647 ))
-    else
-      ISEED=${ISEED:-0}
-    fi
+  if [[ ${DO_OCN_SPPT} = "YES" ]]; then
+    OCNSPPT=${OCNSPPT:-0.8,0.4,0.2,0.08,0.04}
+    OCNSPPT_TAU=${OCNSPPT_TAU:-2.16E4,2.592E5,2.592E6,7.776E6,3.1536E7}
+    OCNSPPT_LSCALE=${OCNSPPT_LSCALE:-500.E3,1000.E3,2000.E3,2000.E3,2000.E3}
+    ISEED_OCNSPPT=$((CDATE*10000 + ${MEMBER#0}*100 + 8)),$((CDATE*10000 + ${MEMBER#0}*100 + 9)),$((CDATE*10000 + ${MEMBER#0}*100 + 10)),$((CDATE*10000 + ${MEMBER#0}*100 + 11)),$((CDATE*10000 + ${MEMBER#0}*100 + 12))
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: please use `current_cycle` in place of `CDATE` here and elsewhere.

File: ush/forecast_postdet.sh [hunk 3]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
   # largest signed integer
-  if [[ "${DO_OCN_SPPT}" = "YES" ]] || [[ "${DO_OCN_PERT_EPBL}" = "YES" ]]; then
-    if [[ ${SET_STP_SEED:-"YES"} = "YES" ]]; then
-      ISEED_OCNSPPT=$(( (current_cycle*1000 + MEMBER*10 + 6) % 2147483647 ))
-      ISEED_EPBL=$(( (current_cycle*1000 + MEMBER*10 + 7) % 2147483647 ))
-    else
-      ISEED=${ISEED:-0}
-    fi
+  if [[ ${DO_OCN_SPPT} = "YES" ]]; then
+    OCNSPPT=${OCNSPPT:-0.8,0.4,0.2,0.08,0.04}
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: should elevate this to `config.efcs` and remove from here.

File: ush/forecast_postdet.sh [hunk 4]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
+    OCNSPPT_TAU=${OCNSPPT_TAU:-2.16E4,2.592E5,2.592E6,7.776E6,3.1536E7}
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: same for this variable.

File: ush/forecast_postdet.sh [hunk 5]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
+    OCNSPPT_LSCALE=${OCNSPPT_LSCALE:-500.E3,1000.E3,2000.E3,2000.E3,2000.E3}
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: and this.

File: ush/forecast_postdet.sh [hunk 6]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
+    ISEED_OCNSPPT=$((CDATE*10000 + ${MEMBER#0}*100 + 8)),$((CDATE*10000 + ${MEMBER#0}*100 + 9)),$((CDATE*10000 + ${MEMBER#0}*100 + 10)),$((CDATE*10000 + ${MEMBER#0}*100 + 11)),$((CDATE*10000 + ${MEMBER#0}*100 + 12))
+  fi
+  if [[ ${DO_OCN_PERT_EPBL} = "YES" ]]; then
+    EPBL=${EPBL:-0.8,0.4,0.2,0.08,0.04}
+    EPBL_TAU=${EPBL_TAU:-2.16E4,2.592E5,2.592E6,7.776E6,3.1536E7}
+    EPBL_LSCALE=${EPBL_LSCALE:-500.E3,1000.E3,2000.E3,2000.E3,2000.E3}
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: raise these to `config.efcs`

File: ush/forecast_postdet.sh [hunk 7]
```diff
@@ -707,13 +712,17 @@ MOM6_postdet() {
...
+    ISEED_EPBL=$((CDATE*10000 + ${MEMBER#0}*100 + 13)),$((CDATE*10000 + ${MEMBER#0}*100 + 14)),$((CDATE*10000 + ${MEMBER#0}*100 + 15)),$((CDATE*10000 + ${MEMBER#0}*100 + 16)),$((CDATE*10000 + ${MEMBER#0}*100 + 17))
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]: please use `current_cycle` in place of `CDATE`.

File: ush/parsing_namelists_FV3.sh [hunk 8]
```diff
@@ -631,6 +631,24 @@ EOF
   use_zmtnblck = ${use_zmtnblck:-".true."}
 EOF
   fi
+  
+  if [[ "${DO_OCN_SPPT:-NO}" == "YES" ]]; then
+    cat >> input.nml <<EOF
+  OCNSPPT=${OCNSPPT:-1.0}
+  OCNSPPT_LSCALE=${OCNSPPT_LSCALE:-500e3}
+  OCNSPPT_TAU=${OCNSPPT_TAU:-21600}
```
Comment (aerorahul) [inline ush/parsing_namelists_FV3.sh]: The `defaults` here will never be realized if they are defined in `config.efcs`, so we should probably set these as:
```suggestion
  OCNSPPT=${OCNSPPT}
  OCNSPPT_LSCALE=${OCNSPPT_LSCALE}
  OCNSPPT_TAU=${OCNSPPT_TAU}
```

File: ush/parsing_namelists_FV3.sh [hunk 9]
```diff
@@ -631,6 +631,24 @@ EOF
...
+  ISEED_OCNSPPT=${ISEED_OCNSPPT:-${ISEED}}
+EOF
+  fi
+
+  if [[ "${DO_OCN_PERT_EPBL:-NO}" == "YES" ]]; then
+    cat >> input.nml <<EOF
+  EPBL=${EPBL:-1.0}
+  EPBL_LSCALE=${EPBL_LSCALE:-500e3}
+  EPBL_TAU=${EPBL_TAU:-21600}
```
Comment (aerorahul) [inline ush/parsing_namelists_FV3.sh]:
```suggestion
  EPBL=${EPBL}
  EPBL_LSCALE=${EPBL_LSCALE}
  EPBL_TAU=${EPBL_TAU}
```

File: ush/parsing_namelists_MOM6.sh [hunk 10]
```diff
@@ -25,24 +25,6 @@ EOF
...
-
-if [[ "${DO_OCN_PERT_EPBL}" == "YES" ]]; then
-  cat >> input.nml <<EOF
-  EPBL=${EPBL:-1.0}
-  EPBL_LSCALE=${EPBL_LSCALE:-500e3}
-  EPBL_TAU=${EPBL_TAU:-21600}
-  ISEED_EPBL=${ISEED_EPBL:-${ISEED}}
-EOF
-fi
-
```
Comment (aerorahul) [inline ush/parsing_namelists_MOM6.sh:45]: I do see that this has been moved to `parsing_namelists_FV3.sh`, but consider the case when we may be running a data atmosphere in which case `parsing_namelists_FV3.sh` will not be called.
Would we then have to move those back from `parsing_namelists_FV3.sh` back to here?

**Review by aerorahul — CHANGES_REQUESTED** [review]
mostly looks great.
some very minor comments and one comment on moving ocean related namelist section out of `parsing_namelists_MOM6.sh` to `parsing_namelists_FV3.sh`.

File: ush/parsing_namelists_MOM6.sh (hunk 10 above)
Comment (NeilBarton-NOAA) [reply ush/parsing_namelists_MOM6.sh:45]: These namelist options are under &nam_stochy, which is defined in the parsing_namlists_FV.sh file.

File: parm/config/gefs/config.base [hunk 11]
```diff
@@ -75,7 +75,7 @@ export NCP="/bin/cp -p"
 export NMV="/bin/mv"
 export NLN="/bin/ln -sf"
 export VERBOSE="YES"
-export KEEPDATA="NO"
```
Comment (WalterKolczynski-NOAA) [inline parm/config/gefs/config.base:77]: What is the reason to template this variable?

File: ush/forecast_postdet.sh [hunk 12]
```diff
@@ -693,6 +693,13 @@ MOM6_postdet() {
       ${NLN} "${COM_OCEAN_ANALYSIS}/${RUN}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/mom6_increment.nc"
   fi
 
+  # GEFS perturbations
+  # TODO if [[ $RUN} == "gefs" ]] block maybe be needed 
+  #     to ensure it does not interfere with the GFS
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: I think it won't be necessary since `ODA_INCUPD` would be False

File: ush/parsing_namelists_MOM6.sh (hunk 10 above)
Comment (aerorahul) [reply ush/parsing_namelists_MOM6.sh:45]: Thanks for explaining @NeilBarton-NOAA

**Review by aerorahul — APPROVED** [review]
Looks good @NeilBarton-NOAA 
Please merge `develop` and resolve the conflicts.
This is ready then

**Comment by NeilBarton-NOAA** [comment]
@aerorahul developed merged

**Comment by WalterKolczynski-NOAA** [comment]
MEMDIR isn't always defined for GFS like it is for GEFS
```
+ forecast_postdet.sh[690]: [[ NO = \Y\E\S ]]
/work2/noaa/stmp/CI/ORION/2385/gfs/ush/forecast_postdet.sh: line 701: MEMDIR: unbound variable
```

File: scripts/exglobal_stage_ic.sh [hunk 13]
```diff
@@ -106,7 +106,7 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
 
     # Ocean Perturbation Files
     # Extra zero on MEMDIR ensure we have a number even if the string is empty
-    if (( 0${MEMDIR:3} > 0 )) && [[ "${OCN_ENS_PERTURB_FILES:-false}" == "true" ]]; then
+    if (( 0${MEMDIR:3} > 0 )) && [[ "${STAGE_OCN_PERTURB_FILES:-false}" == "true" ]]; then
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh]:
```suggestion
    if (( MEMBER > 0 )) && [[ "${STAGE_OCN_PERTURB_FILES:-false}" == "true" ]]; then
```
Use `MEMBER` and update the comment on the previous line.
`MEMBER` is set in https://github.com/NOAA-EMC/global-workflow/blob/0edbdc197441582f9c402da0f3a7f144b88960c7/ush/forecast_predet.sh#L121

File: ush/forecast_postdet.sh [hunk 14]
```diff
@@ -695,6 +695,13 @@ MOM6_postdet() {
...
+  if (( 0${MEMDIR:3} > 0 )) && [[ "${ODA_INCUPD:-False}" == "True" ]]; then
```
Comment (aerorahul) [inline ush/forecast_postdet.sh]:
```suggestion
  if (( MEMBER > 0 )) && [[ "${ODA_INCUPD:-False}" == "True" ]]; then
```

**Review by aerorahul — APPROVED** [review]
looks good.  hopefully this will pass

File: scripts/exglobal_stage_ic.sh [hunk 15]
```diff
@@ -106,7 +106,7 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
 
     # Ocean Perturbation Files
     # Extra zero on MEMDIR ensure we have a number even if the string is empty
-    if (( 0${MEMDIR:3} > 0 )) && [[ "${OCN_ENS_PERTURB_FILES:-false}" == "true" ]]; then
+    if (( MEMBER > 0 )) && [[ "${STAGE_OCN_PERTURB_FILES:-false}" == "true" ]]; then
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh]:
```suggestion
    if (( 0${MEMDIR:3} > 0 )) && [[ "${STAGE_OCN_PERTURB_FILES:-false}" == "true" ]]; then
```
This was my mistake assuming `MEMBER` was defined earlier in this script.

**Comment by WalterKolczynski-NOAA** [comment]
`STAGE_OCN_PERTURB_FILES` is "false" in `config.stage_ic`, so the increment doesn't get copied and then the forecast fails because it can't find the increment.

```
+++ config.stage_ic[9]: export STAGE_OCN_PERTURB_FILES=false
```

```
ls /scratch1/NCEPDEV/global/CI/2385/RUNTESTS/COMROOT/C48_S2SWA_gefs_1b50a39e/gefs.20210323/06/mem001/model_data/ocean/restart          
20210323.120000.MOM.res.nc
```

```
 0: NOTE from PE     0: ==> External_ic::get_nggps_ic: using control file INPUT/gfs_ctrl.nc for NGGPS IC
14: 
14: FATAL from PE     2:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
14: 
18: 
18: FATAL from PE     6:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
18: 
13: 
13: FATAL from PE     1:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
13: 
15: 
15: FATAL from PE     3:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
15: 
17: 
17: FATAL from PE     5:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
17: 
16: 
16: FATAL from PE     4:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
16: 
19: 
19: FATAL from PE     7:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
19: 
12: 
12: FATAL from PE     0:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
12: 
12: 
12: FATAL from PE     0:  initialize_oda_incupd: Unable to open INPUT/mom6_increment.nc
```

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA @aerorahul I edited so the use of ODA in the ocean GEFS is dependent on if those files are staged and re-tested at the CI resolution

**Comment by TerrenceMcGuinness-NOAA** [comment]
The `config.stage_ic` file was not updated. We need to add removing of EXPDIRs as a default behavior.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Restarted CI by first removing all previous runs (e.g.` rm -Rf EXPDIR COMROOT`)

**Comment by aerorahul** [comment]
@bingfu-NOAA 
Would you mind giving this a review?

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by aerorahul** [comment]
@bingfu-NOAA 
Since this PR is trying to replicate part of the EP5* workflow, a review would be great to have.

**Review by bingfu-NOAA — APPROVED** [review]

**Pull request merged**